### PR TITLE
Fix problems related to Onyx.clear and useOnyx

### PR DESCRIPTION
**Pull request by fabioh8010**
### Details

This PR fixes two problems we've discovered on E/App during migration to `useOnyx`, both related to `Onyx.clear`. More details [here](https://github.com/Expensify/App/issues/48725#issuecomment-2400683475).

#### Problem 1

Racing conditions between an `Onyx.clear()` execution and `useOnyx` subscription causes `useOnyx` to be always in `loading` state and never fully resolve its first connection.

This happens because we clear the cache during `Onyx.clear()` execution and if we are connecting to some key with `useOnyx` at the same time, there is no cache where the hook can look into (`OnyxCache.hasCacheForKey(key)` will be `false`) and it get stuck in `loading` state.

This problem doesn't happen for `withOnyx` basically because the racing condition doesn't occur. This can happen by the following reasons:

- As `withOnyx` is a HOC and consequently a wrapper component, it uses React lifecycles / batching mechanisms that will contribute to a slightly rendering/execution delay that is enough to allow `Onyx.clear()` finish its operation.
- The `useOnyx` hook uses `useSyncExternalStore` under the hood, which is designed to not batch its updates and execute and return as fast as possible, thus contributing to the race condition to happen.

#### Problem 2

The cache clearing performed by `Onyx.clear()` combined with the connection reusing logic we currently have causes an `useOnyx` connection to a key that was connected before to be always `loading` state and never fully resolve its first connection.

Specifically, if we are reusing the connection of a key that is still exists in the connection manager after an `Onyx.clear()` execution, the next `useOnyx` to connect to that same key gets stuck in the `loading` state as the cache doesn't really exist anymore, and we didn't populate it again because we are reusing the connection.

When logging out a lot of Onyx connections will be removed from the connection manager as expected, but some of them persists because they don't belong to the React's lifecycle e.g. any `Onyx.connect` in our lib files, helping create the issue described above.

This problem doesn't happen for `withOnyx` because the HOC doesn't reuse connections and create new ones upon every subscription, thus populating the cache again for that key and allowing the HOC to work as expected.

---

This PR also exposes the `reuseConnection` flag in `useOnyx` API for convenience purposes (it's already exposed in `Onyx.connect`).

### Related Issues

https://github.com/Expensify/App/issues/48725

### Automated Tests

Unit tests were added to cover the fixes mentioned.

### Manual Tests

1. Checkout [this branch](https://github.com/fabioh8010/expensify-app/tree/bugfix/48725) on E/App.
2. Link the changes of this Onyx PR branch to E/App.
3. Run `npm test` on E/App and assert all tests are passing.
4. Login with a user, logout and login with the same user again, assert the whole process works as expected.

❗️You will notice that after second login on Web/mWeb, user is redirected to settings page instead of home. Please have in mind that this issue is not related to these fixes, but instead one more fix that would be necessary to be made on E/App when migrating `AuthScreens.tsx` to `useOnyx`.❗️

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/2ec75fd9-2a82-48ec-b900-15d61f0546f4

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/20abd527-ad1f-4c5c-8533-b0e0f2d839ab

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/c9389ba4-f4c6-4ea9-a9be-01adf1e80914

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/858dff39-8952-4cf6-9282-05b8b60fd26e

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/c05274ed-b940-44fa-b5a4-6056564ca7f0

https://github.com/user-attachments/assets/81490232-19a6-4415-b445-61fc2f2b19bf

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/84204411-5c48-45b4-b7d7-3da22f686a49

</details>

### CODE REVIEW

**Comment by hannojg** [comment]
LGTM !

**Review by Beamanator — DISMISSED** [review]
This looks great to me, and looks like it was peer reviewed a few times in Slack 👍 👍 

@tgolen or @marcaaron do either of y'all want to take a look before we merge?

**Review by blazejkustra — APPROVED** [review]

**Comment by fabioh8010** [comment]
Bump @tgolen @marcaaron

**Comment by marcaaron** [comment]
Happy to review this, but I don't yet understand what problem we are solving. Can we update the description with the explanation? I took a look at the comment we linked out to and it says:

>From what I've seen this problem doesn't happen for withOnyx basically because the racing condition doesn't occur (maybe because it's a component and it's slightly slower during initialisation?).

Am I reading that "maybe" correctly? Are we not sure why it happens? Do we have a better explanation somewhere?

File: lib/OnyxConnectionManager.ts [hunk 1]
```diff
@@ -73,12 +73,18 @@ class OnyxConnectionManager {
      */
     private lastCallbackID: number;
 
+    /**
+     * Stores the last generated session ID for the connection manager.
```
Comment (marcaaron) [inline lib/OnyxConnectionManager.ts:77]: Maybe we can add some explanation about what this `sessionID` is used for and why we have it at a ll?

File: lib/OnyxConnectionManager.ts [hunk 2]
```diff
@@ -89,7 +95,10 @@ class OnyxConnectionManager {
      */
     private generateConnectionID<TKey extends OnyxKey>(connectOptions: ConnectOptions<TKey>): string {
         const {key, initWithStoredValues, reuseConnection, waitForCollectionCallback} = connectOptions;
-        let suffix = '';
+
+        // The current session ID is appended to the connection ID so we can have different connections
+        // after an Onyx.clear() operation.
```
Comment (marcaaron) [inline lib/OnyxConnectionManager.ts:100]: >so we can have different connections after an Onyx.clear() operation.

Let's also say why that is important or needed?

File: lib/OnyxConnectionManager.ts [hunk 3]
```diff
@@ -229,6 +238,15 @@ class OnyxConnectionManager {
         this.connectionsMap.clear();
     }
 
+    /**
+     * Refreshes the connection manager's session ID. Used in Onyx.clear() in order to
+     * create new connections after Onyx is cleared, instead of reusing existing connections
+     * that can produce unexpected behavior in Onyx subscribers.
```
Comment (marcaaron) [inline lib/OnyxConnectionManager.ts:244]: Can we get more specific? What is the "unexpected behavior"? I think including an example will help later on if someone needs to make changes here or questions whether it is still required.

File: lib/useOnyx.ts [hunk 4]
```diff
@@ -23,6 +23,12 @@ type BaseUseOnyxOptions = {
      * If set to `true`, data will be retrieved from cache during the first render even if there is a pending merge for the key.
      */
     allowStaleData?: boolean;
+
+    /**
+     * If set to `false`, the connection won't be reused between other subscribers that are listening to the same Onyx key
+     * with the same connect configurations.
+     */
```
Comment (marcaaron) [inline lib/useOnyx.ts:30]: Can we include an example of _when_ might we want to use this?

File: lib/useOnyx.ts [hunk 5]
```diff
@@ -232,8 +238,8 @@ function useOnyx<TKey extends OnyxKey, TReturnValue = OnyxValue<TKey>>(key: TKey
 
         // If the previously cached value is different from the new value, we update both cached value
         // and the result to be returned by the hook.
-        // If the cache was set for the first time, we also update the cached value and the result.
-        const isCacheSetFirstTime = previousValueRef.current === null && hasCacheForKey;
+        // If the cache was set for the first time or we have a pending Onyx.clear() task, we also update the cached value and the result.
+        const isCacheSetFirstTime = previousValueRef.current === null && (hasCacheForKey || OnyxCache.hasPendingTask(TASK.CLEAR));
```
Comment (marcaaron) [inline lib/useOnyx.ts:242]: What does it mean for the cache to be "set first time"? Like, as in the first time the hook runs? It makes sense to consider it "set" when it has cache value. But, can we improve the documentation to explain why we should consider the cache "set" if there is a pending clear task?

**Review by marcaaron — CHANGES_REQUESTED** [review]
Most of my comments are about improving the documentation for now.

File: lib/OnyxConnectionManager.ts (hunk 2 above)
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts]: I moved all explanations to `private sessionID: string;`, please have a look!

File: lib/OnyxConnectionManager.ts (hunk 3 above)
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts]: I moved all explanations to `private sessionID: string;`, please have a look!

File: lib/useOnyx.ts (hunk 5 above)
Comment (fabioh8010) [reply lib/useOnyx.ts]: I refactored the comments, please let me know if it's clearer now.

**Comment by fabioh8010** [comment]
@marcaaron Addressed the comments and added explanations to the PR description, please let me know if it's clearer now.

File: lib/useOnyx.ts (hunk 4 above)
Comment (marcaaron) [reply lib/useOnyx.ts:30]: Ok, I see now. Not really related to these changes.

**Review by marcaaron — APPROVED** [review]
These changes seem reasonable to me.

**Pull request merged**